### PR TITLE
[SPARK-39159][SQL] Add new Dataset API for Offset

### PR DESCRIPTION
**Pull request by beliefer**
### What changes were proposed in this pull request?
Currently, Spark added `Offset` operator.
This PR try to add `offset` API into `Dataset`.

### Why are the changes needed?
`offset` API is very useful and construct test case more easily.

### Does this PR introduce _any_ user-facing change?
'No'.
New feature.

### How was this patch tested?
New tests.

### CODE REVIEW

**Comment by beliefer** [comment]
ping @cloud-fan

File: sql/core/src/test/scala/org/apache/spark/sql/DataFrameSuite.scala [hunk 1]
```diff
@@ -605,6 +605,20 @@ class DataFrameSuite extends QueryTest
...
+      testData.offset(90),
+      testData.collect().drop(90).toSeq)
+
+    checkAnswer(
+      arrayData.toDF().offset(99),
+      arrayData.collect().drop(99).map(r => Row.fromSeq(r.productIterator.toSeq)))
+
+    checkAnswer(
+      mapData.toDF().offset(99),
+      mapData.collect().drop(99).map(r => Row.fromSeq(r.productIterator.toSeq)))
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/DataFrameSuite.scala:619]: can we test a few combinations? like limit followed by offset, and offset followed by limit.

**Review by cloud-fan — APPROVED** [review]

**Review by cloud-fan — APPROVED** [review]

File: sql/core/src/test/scala/org/apache/spark/sql/DataFrameSuite.scala (hunk 1 above)
Comment (beliefer) [reply sql/core/src/test/scala/org/apache/spark/sql/DataFrameSuite.scala:619]: OK

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Comment by beliefer** [comment]
@cloud-fan Thank you!

File: sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala [hunk 2]
```diff
@@ -2102,6 +2102,16 @@ class Dataset[T] private[sql](
     Limit(Literal(n), logicalPlan)
   }
 
+  /**
+   * Returns a new Dataset by skipping the first `m` rows.
```
Comment (jerqi) [inline sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala:2106]: a typo? may be 'n'?

Comment (beliefer) [reply sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala:2106]: Yeah. this is a minor issue. I will fix it by the way.

**Pull request closed without merging**